ntp-proto/src/filter.rs

```diff
@@ -451,6 +444,116 @@ fn find_interval(chime_list: &[CandidateTuple]) -> Option<(NtpDuration, NtpDurat
+
+    let mut qmax_index = 0;
+
+    let mut min_peer_jitter = 2.0e9;
+    let mut max_selection_jitter = -2.0e9;
+
+    loop {
+        // for each candidate
+        for (index, candidate) in candidates.iter().enumerate() {
+            let p = candidate.peer;
+
+            min_peer_jitter = p.statistics.jitter.min(min_peer_jitter);
+
+            // compute the selection jitter
+            let selection_jitter = candidates
+                .iter()
+                .map(|q| p.statistics.offset - q.peer.statistics.offset)
+                .map(|v| v.to_seconds().powi(2))
+                .sum::<f64>()
+                .sqrt();
+
+            if selection_jitter > max_selection_jitter {
+                qmax_index = index;
+                max_selection_jitter = selection_jitter;
+            }
+        }
+
+        // If the maximum selection jitter is less than the
+        // minimum peer jitter, then tossing out more survivors
+        // will not lower the minimum peer jitter, so we might
+        // as well stop.  To make sure a few survivors are left
+        // for the clustering algorithm to chew on, we also stop
+        // if the number of survivors is less than or equal to
+        // NMIN (3).
+        if max_selection_jitter < min_peer_jitter || candidates.len() <= MIN_CLUSTER_SURVIVORS {
+            return;
+        }
+
+        // delete the survivor qmax (the one with the highest jitter) and go around again
+        candidates.remove(qmax_index);
+    }
+}
+
+#[allow(dead_code)]
+fn clock_select(
+    peers: &[Peer],
+    local_clock_time: NtpTimestamp,
+    system_poll: NtpDuration,
+) -> Option<Vec<SurvivorTuple>> {
+    let valid_associations = peers
+        .iter()
+        .filter(|p| p.accept_synchronization(local_clock_time, system_poll));
+
+    let candidates = construct_candidate_list(valid_associations, local_clock_time);
+
+    let mut survivors = construct_survivors(&candidates, local_clock_time);
+
+    if survivors.len() < MIN_INTERSECTION_SURVIVORS {
+        return None;
+    }
+
+    cluster_algorithm(&mut survivors);
+
+    Some(survivors)
+}
+
+#[allow(dead_code)]
+struct ClockCombine {
+    offset: NtpDuration,
+    jitter: NtpDuration,
+}
+
+/// Combine the offsets of the clustering algorithm survivors
+/// using a weighted average with weight determined by the root
+/// distance.  Compute the selection jitter as the weighted RMS
+/// difference between the first survivor and the remaining
+/// survivors.  In some cases, the inherent clock jitter can be
+/// reduced by not using this algorithm, especially when frequent
+/// clockhopping is involved.  The reference implementation can
+/// be configured to avoid this algorithm by designating a
+/// preferred peer.
+#[allow(dead_code)]
+fn clock_combine<'a>(
+    survivors: &'a [SurvivorTuple<'a>],
+    local_clock_time: NtpTimestamp,
+) -> ClockCombine {
+    let mut y = 0.0;
+    let mut z = 0.0;
+    let mut w = 0.0;
```

Comment:
What are these?

Reply:
I added some comments. But I have to say the spec is particularly vague here 

> The routine processes peer offset and jitter statistics to produce the combined system offset THETA and system peer jitter PSI_p, where each server statistic is weighted by the reciprocal of the root synchronization distance and the result normalized.

jitter is calculated from the offsets in the code skeleton, but I honestly don't understand why that is, and why the peer's `jitter` is not considered.

Reply:
Ok, this might be a bug in the code skeleton then.

Reply:
yeah, it says

```
        /*
         * Combine the offsets of the clustering algorithm survivors
         * using a weighted average with weight determined by the root
         * distance.  Compute the selection jitter as the weighted RMS
         * difference between the first survivor and the remaining
         * survivors.  In some cases, the inherent clock jitter can be
         * reduced by not using this algorithm, especially when frequent
         * clockhopping is involved.  The reference implementation can
         * be configured to avoid this algorithm by designating a
         * preferred peer.
         */
        y = z = w = 0;
        for (i = 0; s.v[i].p != NULL; i++) {
                p = s.v[i].p;
                x = root_dist(p);
                y += 1 / x;
                z += p->offset / x;
                w += SQUARE(p->offset - s.v[0].p->offset) / x;
        }
        s.offset = z / y;
        s.jitter = SQRT(w / y);
}
```
so we should turn that into ` w += SQUARE(p->jitter - s.v[0].p->jitter) / x;` then?

Reply:
No, i think the above just recalculates PSI_s, and we should afterwards combine that with PSI_p per the formula on page 43 to get the total jitter.

Reply:
most likely, we're both clicking on a link in this comment to view the code, and that takes you to the now-outdated version. Very annoying... So you have to navigate via "files changed"

Reply:
Sorry, with the above I meant the excerpt of c code you posted. From how I understand it, that just recalculates the selection jitter in w (or at least the formulas seem very similar). From what I understand, PSI_p is just the peer jitter from the peer that we selected as the system peer, that is, the peer which came first in the list after clock_select.

The alternative you suggest (w += SQUARE(p->jitter - s.v[0].p->jitter) / x) just seems downright weird to me, as that is just taking the squared difference of two jitters, which I am not sure what that is. (but probably not a jitter)

Reply:
yes, not sure why but the skeleton decides to calculate just PSI_p in that function. The current code of this PR calculates the full system jitter (PSI) now.